File: backend/app/services/units.py
"""
Shared, dependency-free unit conversion. Lives on its own so that
constraint_engine.py and layout_renderer.py don't need to import each
other just to agree on how a room's dimensions convert to centimetres
-- each pipeline stage should only ever depend on schemas/, never on
another stage's module.
"""

from app.schemas.room import RoomDimensions

_UNIT_TO_CM = {"cm": 1.0, "m": 100.0, "ft": 30.48}


def room_dimensions_cm(dimensions: RoomDimensions) -> tuple[float, float]:
    """Returns (length_cm, width_cm), converting from whatever unit
    the room was recorded in (RoomDimensions.unit defaults to 'ft',
    while every FurnitureItem/Placement dimension is in cm)."""
    factor = _UNIT_TO_CM.get(dimensions.unit, 1.0)
    return dimensions.length * factor, dimensions.width * factor

File: frontend/src/lib/units.ts
import type { RoomDimensions } from "../types";

const UNIT_TO_CM: Record<string, number> = { cm: 1, m: 100, ft: 30.48 };

/** Mirrors backend/app/services/units.py::room_dimensions_cm */
export function roomDimensionsCm(dimensions: RoomDimensions): [number, number] {
  const factor = UNIT_TO_CM[dimensions.unit] ?? 1;
  return [dimensions.length * factor, dimensions.width * factor];
}

export function cmToUnit(cm: number, unit: string): number {
  const factor = UNIT_TO_CM[unit] ?? 1;
  return cm / factor;
}

export function formatCm(cm: number): string {
  return `${Math.round(cm)} cm`;
}
